Improvement/dtm guide (#473)
SoftwareAG/c8y-docs: content/dtm/asset-types-bundle/property-library.md

---
weight: 10
title: Property library
layout: redirect
---

> 

Author: No, they do not refer to the same thing. Custom properties are a part of the Assets created by user. First the custom properties are created. Later assets are created using these custom properties.

> **Property library** section allows users to create Custom properties. Custom properties are the attributes that define an asset type. One or more custom properties are used to define an asset type.

Reviewer: **Property library** section > The **Property library** section

> **Property library** section allows users to create Custom properties. Custom properties are the attributes that define an asset type. One or more custom properties are used to define an asset type.

Author: Done

Also, the same custom property can be part of 2 different asset types.

> When user navigates to Property library page via the path “Configuration> Digital twin library > Property library” for the first time, there will be no custom property present in the system. User can add custom properties using **Add custom property** at the bottom of the screen.

Author: Converted to direct speech. Will convert the path to link once all chapters are available.

> When user navigates to Property library page via the path “Configuration> Digital twin library > Property library” for the first time, there will be no custom property present in the system. User can add custom properties using **Add custom property** at the bottom of the screen.

Author: Done.

> Screenshot to be added when no custom properties are present.

Author: Done.

> When user logs in to a tenant where the custom properties are already present, then the list of existing custom properties are displayed on the left part of screen. On the right side of screen, user is prompted to add or select a custom property.

Author: Done.

Screenshot to be added

On click of **Add custom property** a screen is displayed on right side. Once the user fills in all the mandatory fields with valid inputs, then user has an option to save the custom property.
Screenshot to be added for Add custom property screen

Custom properties have the following attributes:

> *	Label: Name of the Custom property and is a mandatory field. Custom property is addressed using the Label field.

Reviewer: I would actually better make this a table.

*	Key: Mandatory field Key, is used to uniquely identify and store the custom property internally.
*	Description: Optional field description is used to provide a brief description about the custom property being created.

> *	Complex property toggle: This toggle option when enabled creates a Complex custom property. By default its disabled and creates a simple custom property.

Reviewer 2: ```suggestion
*	**Complex property toggle**:  This information is optional. Use this toggle to create a complex custom property, that contains one or multiple key-value pairs. The toggle is disabled by default.
```

*	Type: Type dropdown defines the type of the custom property/ attribute. Type can belong to any of these categories – Text, Number, File upload, Date picker, Enumeration or Boolean.

> *	Default value: Optional text box default value, defines the default value for the attribute when filled. If empty, user has an option to fill, and if filled then user can modify default value during creation of Asset.

Reviewer 2: Format attribute in bold.
The spelling in the UI needs to be adjusted to "Default value" (currently "Default Value").

"Optional text box default value, defines the default value for the attribute when filled. If empty, user has an option to fill, and if filled then user can modify default value during creation of Asset."

--> This information is optional. It defines the default value for the respective attribute. You can enter any desired information. If this field already contains information, you can modify it during the creating of an asset.

Question: How is this linked to the creation of an asset? Add the respective link here.

> >**Info:** Default value text box gets displayed only if the ‘Type’ chosen is either **Text** or **Number**.

Reviewer 2: ">**Info:** Default value text box gets displayed only if the ‘Type’ chosen is either **Text** or **Number**."

--> {{< c8y-admon-info>}}
The **Default Value** field is only visible if you select either "Text" or "Number". 
{{< /c8y-admon-important >}}

Reviewer 2: @nupadhyaya Please add this change

> 

Reviewer 2: Add a level 4 (####) heading here, for example "#### Custom property types"

Reviewer 2: @nupadhyaya Please add this change

> As mentioned above, user can choose any of the 6 below mentioned options for **Type** dropdown:

Reviewer 2: "As mentioned above, user can choose any of the 6 below mentioned options for **Type** dropdown:"

--> As described above, you can select one of the six options listed under **Type** to further define the type of your custom property.

Author: Done

> 1. **Text**: In type dropdown when **Text** is selected, then 4 more optional fields appear in **Add new property** page layout. **Text** is chosen when the custom property value will be a string.

Reviewer 2: "1. **Text**: In type dropdown when **Text** is selected, then 4 more optional fields appear in **Add new property** page layout. **Text** is chosen when the custom property value will be a string."

-->Make "Text" a level 5 heading (##### Text).

If you select "Text", four additional fields appear. The information in these fields is optional. Select "Text" when the custom property value is a string, for example, the building name.

Author: Done

>   >**Example**: Building Name.

Reviewer 2: Remove, see comment above

Author: Done.

  Screenshot to be added by choosing Text in dropdown

>   *	**Min Length**: When the checkbox against this field is checked, user has to enter the minimum length of the text to be provided for this Custom property during Asset creation. Any text below the minimum length will not be accepted.

Reviewer 2: "**Min Length**: When the checkbox against this field is checked, user has to enter the minimum length of the text to be provided for this Custom property during Asset creation. Any text below the minimum length will not be accepted."

--> Make "Min length" a level 5 heading (see above). 

If you check this box, you need to enter the minimum length of the text that must be provided for this custom property during the asset creation. Any text below the minimum length will not be accepted."

Note that the spelling of "Length" needs to be changed in the UI to "length".

Reviewer: If you check this box > Better: If selected,

>     User also has an option of not checking the check box, which implies there will be no minimum length criteria for the custom property value during asset creation.

Reviewer 2: "User also has an option of not checking the check box, which implies there will be no minimum length criteria for the custom property value during asset creation."

--> You also have the option of not checking this box. Doing so implies that there will be no minimum length criteria set for this custom property during the asset creation.

Reviewer: You also have the option of not checking this box. > Better: If not selected, ...

Author: Done.

>   * **Max Length**: If the check box against this field is checked, then user has to provide the maximum length of text that can be provided as custom property value when creating an asset.

Reviewer: Any text above the maximum length will not be accepted. > This is already mentioned in the next sentence.

>   * **Max Length**: If the check box against this field is checked, then user has to provide the maximum length of text that can be provided as custom property value when creating an asset.

Author: Done

>     Incase the text provided by user exceeds the maximum length, then user is notified regarding the same and asset will not be created.

Author: Done

>   * **RegExp**: This field when checked expects a valid regular expression as input. Based on the regular expression field defined here, the Custom property value has to be provided when creating the asset.

Author: Modified the sentence slightly : 
  If you check this box, you need to enter a valid regular expression. You also need to provide the custom property value when creating the asset, which adhers to the regular expression defined.

>   >**Example**: Incase “Employee ID” is the custom property and all the employee IDs have to start with prefix “SAGEMPID” then the regular expression can be provided as ^SAGEMPID[a=zA-Z0-9]*$

Author: Done

>   *	**Default value**: This optional text box provides the flexibility of assigning a default value to the custom property, if provided during creation of custom property. Incase the default value is left blank, then user can fill in a value for this custom property during asset creation.

Author: Done

>   >**Note:** Default value filled by the user has to confine to all the validations provided by the user. If user has mentioned inputs for Min length, max length and/ or regular expression fields, then the default value should fulfil all the validations.

Author: Done

>   >**Example**: If Employee ID is the custom property being created, and Min length field is 5, Max length field is 20 and RegExp field is defined as ^SAG[a=zA-Z0-9]*$, then the default value should fulfil all the 3 criteria. Failure to do so will result in the validation error and the respective field will be highlighted.

Reviewer 2: "**Example**: If Employee ID is the custom property being created, and Min length field is 5, Max length field is 20 and RegExp field is defined as ^SAG[a=zA-Z0-9]*$, then the default value should fulfil all the 3 criteria. Failure to do so will result in the validation error and the respective field will be highlighted."

-->
**Example:** If the custom property is "Employee ID" with a **min length** of 5, a **max length** of 20 and the **regular expression** defined as "^SAG[a=zA-Z0-9]*$", the default value must fulfill all of these three criteria. If it does not this results in a validation error. The respective field will be highlighted to modify its content to fit the criteria.

   [Screenshot to be added for validation error.]

>   Incase the user provides a value like “SAGEMPID19087” which fulfils all the 3 conditions, then it’s a valid input for the custom property and it is saved with this default value.

Reviewer 2: If you enter the information correctly according to the set criteria, the default value information for this custom property is valid. Click **Save** to finish.

Author: Done

>   [Screenshot to be added for this positive flow.]

Reviewer 2: No screenshot necessary.

Author: Done

> 2. **Number**: If in **Type** dropdown, **Number** is chosen as type, then during asset creation user has to provide a numeric value for this custom property.

Reviewer 2: Make "Number" a level 5 ##### heading. 

If in **Type** dropdown, **Number** is chosen as type, then during asset creation user has to provide a numeric value for this custom property.

--> If you select the **Number** as the custom property type you have to provide a numeric value when creating the asset. 
+ combine with the example below
--> If you select the **Number** as the custom property type you have to provide a numeric value when creating the asset, for example, the height of a building.

Author: Done

>   >**Example**: Building height is a custom property whose value will be numeric.

Reviewer 2: combine with text above (see comment above)

Author: Done

>   When Number is chosen as type, below 3 optional fields appear on screen:

Reviewer 2: The type **Number** contains the following fields for additional information:

Author: Done

>   * **Minimum**: If the check box against this field is checked, then user has to provide a number in the textbox. When creating an asset, this Custom property value should not be lesser than the Minimum value provided during creation of Custom property.

Reviewer 2: format **Minimum** in bold.

If the check box against this field is checked, then user has to provide a number in the textbox. When creating an asset, this Custom property value should not be lesser than the Minimum value provided during creation of Custom property.

--> If you select this box, you must enter a number in the field on the right. When you create the asset, the custom property value cannot be lower or less than the **Minimum** provided.

Author: Done

>    Also, in case user provides a default value during creation of custom property, the default value should be greater than or equal to the minimum value provided.

Reviewer 2: Also, in case user provides a default value during creation of custom property, the default value should be greater than or equal to the minimum value provided.

--> format this as an Info box.

{{< c8y-admon-info>}}
If you have entered a (**Default value**)[add link] beforehand, the **Default value** needs to be greater than or equal to the **Minimum** entered here. 
{{< /c8y-admon-info>}}

Author: Done. Link to Default value need to be added.

>    By default, this checkbox is not checked, and user can choose not to enter the minimum value for the custom property.

Reviewer 2: combine with the info box above.

Author: Done

>   * **Maximum**: If the check box against this field is enabled by user, then a number has to be provided and the custom property value cannot exceed this number.

Author: Done

>    In case default value is provided when creating the custom property, then it has to be lesser than or equal to the maximum value entered by user.

Author: Done

   By default, the check box is not checked, and it is an optional field.

>    * **Default value**: Is an optional field and if filled, then during asset creation the custom property will be initialized with this value. User has an option to modify the default value and enter a new value during asset creation.

Reviewer 2: format **Default value** in bold. Add line breaks.

Is an optional field and if filled, then during asset creation the custom property will be initialized with this value. User has an option to modify the default value and enter a new value during asset creation.

-->  This field is optional. When creating the asset, the custom property is initialized using this value. During the asset creation you have the option to change the **Default value** or enter a new value altogether.

>    If the minimum and/ or maximum fields are entered, then the default value provided will be validated against one or both the conditions.

Reviewer 2: If the minimum and/ or maximum fields are entered, then the default value provided will be validated against one or both the conditions.

--> format this as an info box.

{{< c8y-admon-info>}}
If the **Minimum** and/or **Maximum** is checked, the given **Default value** is validated against one or both of these values.

>    >**Example**: If Building height is a custom property being created, and Minimum field is set to 80 and Maximum is set to 150, then the default value should be between 80 to 150 (Inclusive).

Reviewer 2: **Example**: If Building height is a custom property being created, and Minimum field is set to 80 and Maximum is set to 150, then the default value should be between 80 to 150 (Inclusive).

- format colon (:) in bold
--> If the custom property is the building's height with the **Minimum** field set to 80 and the **Maximum** set to 150, the default value must be between 80 to 150.

Author: Done

>    >**i.e,** Minimum >= default value <= Maximum.

Reviewer 2: remove this content

Author: Done

>   [Screenshot to be added when type is number and all fields filled.]

Reviewer 2: no screenshot necessary.

Author: Done

> 3. **Date Picker**: In case user chooses this option in the **Type** dropdown, then during asset creation for this custom property a date picker/ calendar will be displayed. User must choose a date and custom property will be initialized to that value.

Author: Modified the sentence slightly. Please review.

>   >**Example**: If **Start date** is the custom property, then it can be of type **Date Picker** and when asset is being created, user can initialize this custom property with the relevant start date.

Author: Done

>   [Screenshot to be added for date picker option in custom property screen and during asset creation.]

Reviewer 2: No calendar option. Put this with asset creation.

> 4.	**Enumeration**: If the custom property value has to be chosen from a set of predefined options, then the **Type** can be chosen as **Enumeration** when creating the custom property.

Reviewer 2: re-write sentence to portray information better

> 4.	**Enumeration**: If the custom property value has to be chosen from a set of predefined options, then the **Type** can be chosen as **Enumeration** when creating the custom property.

Author: Modified the content, please review.

>   On choosing this as **Type**, “Values (comma separated)” text box appears on screen. User has to enter all the options separated by comma in this text box.

Author: Done

  When the asset is being created, then these values appear as menu options in the dropdown. User must select one of the options and custom property is initialized with that option.

  >**Example**: If **building color** is a custom property and the color has to be either "black", "white" or "grey", then these values are provided as options to user when creating the custom property.

  [Screenshot to be added for custom property screen and during asset creation.]

5. **Boolean**: **Type** is chosen as **Boolean** when presence of the Custom property can be either "true" or "false".

  >**Example**: If **Whiteboard** is a custom property and in a building there are rooms with Whiteboard and other rooms not having the Whiteboard, then in such cases the custom property is created with type as **Boolean**.

>  When the asset is being created, if the room is having a **Whiteboard**, then it will be initialized to "true" and it will be initialized to "false" in case the Whiteboard is absent.

Reviewer 2: asset creation put to asset creation

  [Screenshot to be added for custom property and asset creation part.]

 6. **File Upload**: This option will be chosen in **Type** dropdown, when a file has to be uploaded for the custom property being defined.

 When the **Type** is chosen as **File Upload**, then user has to mention the allowed file types separated by comma in the “Allowed file types” text box. When creating the asset, files belonging to only these categories will be allowed to be uploaded for the custom property.

 >**Example**: In case user wants to upload the blueprint for a building, then a custom property named “Building blueprint” can be created. When the asset is being created, then the blueprint can be uploaded as a file for this custom property. Once the asset is created, user will have an option to view this file for future reference.  

  [Screenshot to be added for custom property and asset creation part.]

>   Custom properties are of 2 types.

Reviewer 2: @nupadhyaya Please edit content according to this comment.

> * **Simple Custom property**:

Reviewer 2: @nupadhyaya Please edit content according to this comment.

> If an attribute consists of a single key value pair, then it is a Simple custom property.

Reviewer 2: @nupadhyaya Please edit content according to this comment.

 >**Example**: If “Building Color” is defined as a custom property under Property library section, then “building_color” is the key used. “White” is the default value for this custom property.

> [Screenshot to be provided]

Reviewer 2: @nupadhyaya Please edit content according to this comment.

> * **Complex Custom property**:

Reviewer 2: @nupadhyaya Please edit content according to this comment.

If a Custom property consists of multiple key value pairs, then such properties are Complex custom properties. In this case, there will be a root object followed by one or more key value pairs.

>  >**Example**: If Building dimension is defined as a complex custom property, then building area, building height and building width are the multiple key value pairs defining the attribute building dimension.

Reviewer: If the custom property is named "Building dimensions",  > @MWindel This is not the same sense. I think it must stay "If "Building dimension" is defined as a complex custom property ...

[Screenshot to be provided]

Once user has entered all the mandatory fields with valid inputs, then custom property will be created on click of **Save** button. Either a Simple or a Complex custom property can be created, based on inputs provided by user. In case of Complex custom property, each key value pair must be defined, and type should be provided for each key value pair.

On successful creation, user will be notified with a green pop-up notification on top right of screen.

In case the mandatory fields are not filled, then the Save button is not enabled.
If the mandatory fields are filled, but there are validation issues, then the same will be highlighted with red highlight on the corresponding element and validation error text will be displayed on screen. User can then change the inputs fields accordingly and save with new valid inputs.

> [Screenshot to be added for validation issue.]

Author: Will add screenshot

> In case user wants to cancel the creation of custom property, then the same is done by clicking on **Cancel** option on the bottom. No pop-up notification is displayed in such cases and **Add new property** screen gets removed from the RHS pane.

Author: Done

Once the custom properties are created, then all the custom properties get listed on the screen towards the left. In case of simple custom properties, the **type** gets displayed along with the labels. In case of complex properties, it displays **Object** against the labels.

> User has an option to select any of the custom properties. On click, the details get displayed on the right. User can either view or edit any of the fields and save the new values using the **Save** option. On successful edit, user is notified with a success pop-up notification indicating the changes are saved successfully.

Reviewer: I would remove the last sentence. This is obvious.

> User has an option to select any of the custom properties. On click, the details get displayed on the right. User can either view or edit any of the fields and save the new values using the **Save** option. On successful edit, user is notified with a success pop-up notification indicating the changes are saved successfully.

Author: Done

> [Screenshot to be added for Modify scenario.]

Reviewer 2: this screenshot needs to be discussed if necessary or not.

Author: Done. Removed screenshot placeholder.

> User also has an option to delete the created custom property. Click on any custom property and details get displayed on right. Click the **Delete** option on bottom. Custom property gets deleted, and user is notified about the same with a success pop-up notification.

Reviewer: See above

> User also has an option to delete the created custom property. Click on any custom property and details get displayed on right. Click the **Delete** option on bottom. Custom property gets deleted, and user is notified about the same with a success pop-up notification.

Author: Done

>  >**Note**: In case the custom property is being used as part of an existing asset, then the property gets deleted from Asset type and existing Assets.

Author: Done

> Additionally in Property library page, there is a **Search** option provided. In case user wants to search for a particular custom property, then the label can be entered in the search bar and user can click on the search icon.

Author: Done. Search is a magnifying glass.

> Successful search will display the custom properties filtered based on the search text.

Author: Done

> In case there are no custom properties matching the search text, then same is notified to the user with a message on screen.

Author: Done

> Once user defines all the custom properties, then next step is to start defining the Asset types.

Author: Asset type section is already present separately. Modified the statement.

---
weight: 10
title: Property library
layout: redirect
---

**Property library** section allows you to create custom properties. Custom properties are the attributes that define an asset type. One or more custom properties are used to define an asset type.
Also, the same custom property can belong to 2 different asset types.

When you navigate to Property library page via the path **“Configuration> Digital twin library > Property library”** for the first time, there will be no custom property present in the system. Click **Add custom property** to add the custom properties.

If you use a tenant with custom properties already defined in the Property library, you can find the list of defined custom properties on the left, as well as the selected custom properties listed on the right.

On click of **Add custom property** a screen is displayed on right side. Once you fill in all the mandatory fields with valid inputs, then you have an option to save the custom property.
Screenshot to be added for Add custom property screen

> Custom properties have the following attributes:

Reviewer 2: Make this an HTML table.

*	Label: Name of the custom property and is a mandatory field. Custom property is addressed using the Label field.
*	Key: Mandatory field Key, is used to uniquely identify and store the custom property internally.
*	Description: Optional field description is used to provide a brief description about the custom property being created.
*	Complex property toggle: This toggle option when enabled creates a Complex custom property. By default its disabled and creates a simple custom property.

> *	Type: Type dropdown defines the type of the custom property. Type can belong to any of these categories – Text, Number, File upload, Date picker, Enumeration or Boolean.

Reviewer 2: ```suggestion
*	**Type**: This field defines the type of the custom property. It can be any of the following categories – Text, Number, File upload, Date picker, Enumeration or Boolean.
```

> *	Default value: Optional text box default value, defines the default value for the attribute when filled. If empty, then you have an option to fill, and if filled then you can modify default value during creation of Asset.

Reviewer 2: ```suggestion
*	**Default value**: This information is optional. It defines the default value for the attribute. You can modify the content when creating an asset using this custom property.
```

>**Info:** Default value text box gets displayed only if the ‘Type’ chosen is either **Text** or **Number**.

As described above, you can select one of the six options listed under **Type** to further define the type of your custom property.

> ##### **Text**

Reviewer 2: ```suggestion
##### Text
```

  If you select "Text", four additional fields appear. The information in these fields is optional. Select "Text" when the custom property value is a string, for example, the building name.

>   Screenshot to be added by choosing Text in dropdown

Reviewer 2: @nupadhyaya add screenshot

  *	**Min Length**:

>   If you check this box, you need to enter the minimum length of the text that must be provided for this custom property during the asset creation. Any text below the minimum length will not be accepted.

Reviewer 2: ```suggestion
  If selected, you need to enter the minimum length of the text that must be provided for this custom property during the asset creation. Any text below the minimum length will not be accepted.
```

>   You also have an option of not checking this box. Doing so implies there will be no minimum length criteria set for this custom property during the asset creation.

Reviewer 2: ```suggestion
If not selected, there will be no minimum length criteria set for this custom property during the asset creation.
```

  * **Max Length**:

>   If you check this box, you need to enter the maximum length of the text that must be provided for this custom property during the asset creation. Any text above the maximum length will not be accepted.

Reviewer 2: ```suggestion
  If selected, you need to enter the maximum length of the text that must be provided for this custom property during the asset creation.
```

  If the entered text exceeds the maximum length, you get notified of the incidence. The asset will not be created.

  * **RegExp**:

>   If you check this box, you need to enter a valid regular expression. You also need to provide the custom property value when creating the asset, which adhers to the regular expression defined.

Reviewer 2: ```suggestion
  If selected, you need to enter a valid regular expression. You also need to provide the custom property value when creating the asset, which adhers to the regular expression defined.
```

   By default, this box is not checked.

   {{< /c8y-admon-info>}}

  * **Maximum**:

   This value is optional. If you select this box, you must enter a number in the field on the right. When you create the asset, the custom property value cannot exceed the **Maximum** provided.

   {{< c8y-admon-info>}}

   If you have entered a (Default value)[add link] beforehand, the **Default value** needs to be lesser than or equal to the **Maximum** entered here.

   By default, this box is not checked.

   {{< /c8y-admon-info>}}

   * **Default value**:

   This field is optional. When creating the asset, the custom property is initialized using this value. During the asset creation you have the option to change the **Default value** or enter a new value altogether.

   {{< c8y-admon-info>}}

   If the **Minimum** and/or **Maximum** is checked, the given **Default value** is validated against one or both of these values.

   {{< /c8y-admon-info>}}

   **Example:**

   If the custom property is the building's height with the **Minimum** field set to 80 and the **Maximum** set to 150, the **Default value** must be between 80 to 150.

##### **Date Picker**

>   If you select this option in the **Type** dropdown list, then during asset creation you must choose a date and custom property will be initialized to that value.

Reviewer 2: ```suggestion
  If you select this option in the **Type** drop-down list, you must choose a date and custom property during asset creation to be initialized to that value.
```

  **Example:**

   An example for a custom property with this type is "start date". During the asset creation you can initialize the custom property with the relevant start date.

##### **Enumeration**

>   If you want to choose a value from a dropdown during asset creation, then select **Type** as **Enumeration** when creating the custom property.

Reviewer 2: ```suggestion
  If you want to choose a value from the drop-down menu during the asset creation, then select **Type** as **Enumeration** when creating the custom property.
```

>   When you select **Enumeration** as the custom property **Type**, a second field appears below. Here, enter all desired information separated by a comma.

Reviewer 2: Combine into one paragraph with the sentence above.

>   When the asset is being created, then these values appear as menu options in the dropdown. Select one of the options and custom property is initialized with that option.

Reviewer 2: ```suggestion
   During asset creating, these values appear as menu options in the drop-down menu. Select one of the options to initialize the custom property with that option.
```

Reviewer 2: Also, combine with paragraph/sentences above

  **Example:**

>    If "building color" is a custom property and the color has to be either "black", "white" or "grey", then these values are provided as comma separated options when creating the custom property.

Reviewer 2: ```suggestion
   If you create a custom property for "building color" and the color must be either black, white or grey, you have to provide these options separated by a colon when creating the custom property.
```

>   [Screenshot to be added for custom property screen.]

Reviewer 2: remove

##### **Boolean**

>   If value of the Custom property is either "true" or "false", then choose **Type** as **Boolean**.

Reviewer 2: ```suggestion
  If you select Boolean for your custom property type, the presence of the custom property can either be true or false.
```

  **Example:**

>   If "Whiteboard" is a custom property and in a building there are rooms with Whiteboard and other rooms not having the Whiteboard, then in such cases the custom property is created with type as **Boolean**.

Reviewer 2: ```suggestion
  If you define a custom property called "Whiteboard" for a building where there are rooms with a whiteboard and some without, then the custom property needs to be created as Boolean. So when the asset is created, if the room has a whiteboard, it is then initialized to "true". If there is no whiteboard in the room, the asset is initialized to "false".
```

>   Initialize it to "true" if whiteboard is present in room, and initialize it to "false" in case the whiteboard is absent.

Reviewer 2: ```suggestion
  Initialize it to "true" if a whiteboard is present in the room, and initialize it to "false" if there is no whiteboard.
```

> ##### **File Upload**

Reviewer 2: ```suggestion
##### **File upload**
```

>   If you want to upload a file when creating the asset, then choose **File Upload** as the **Type**.

Reviewer 2: ```suggestion
  Select this option if you need to upload a file to define the custom property.
```

Reviewer 2: @nupadhyaya Please check with the unification of the UI from capital spelling to "File upload"

>   On selecting **File Upload** additional text box is displayed to mention the allowed file types. This is an optional field and if filled, only the mentioned file types are allowed to be uploaded.

Reviewer 2: ```suggestion
  If you select **File upload**, you see an additional text box showing the allowed file types. Enter the allowed file types separated by a comma in the **Allowed file types** field. During the asset creation, you can only upload files of the allowed file types.
```

Reviewer 2: @nupadhyaya please check if content is correct

 **Example:**

>   In case you want to upload the blueprint for a building, then create a custom property named “Building blueprint”. When creating the asset, the blueprint can be uploaded as a file for this custom property. Once the asset is created, you will have an option to view this file for future reference.  

Reviewer 2: ```suggestion
  If you want to upload the blueprint for a building,you must name the custom property "Building blueprint" and specify the file type of the blueprint file. When you create the asset, you can upload the respective file. Once the asset is created, you can view this file for future reference.
```

  Custom properties are of 2 types.

* **Simple Custom property**:

If an attribute consists of a single key value pair, then it is a Simple custom property.

 **Example:**

>  If “Building Color” is defined as a custom property under Property library section, then “building_color” is the key used. “White” is the default value for this custom property.

Reviewer 2: ```suggestion
 If the custom property is named "Building color", you must use the key "building_color" with the Default value being "white".
```

[Screenshot to be provided]

* **Complex Custom property**:

>  If a Custom property consists of multiple key value pairs, then such properties are Complex custom properties. In this case, there will be a root object followed by one or more key value pairs.

Reviewer 2: ```suggestion
 A complex custom property consists of multiple key value pairs, meaning it contains a root object followed by one or more key value pairs. You can select the complex custom property option by turning on the toggle. If you turn on the toggle, you must define each key value pair as well as providing a Type for each pair.

In case of Complex custom property, each key value pair must be defined, and type should be provided for each key value pair
```

 **Example:**

>  If Building dimension is defined as a complex custom property, then building area, building height and building width are the multiple key value pairs defining the attribute building dimension.

Reviewer 2: ```suggestion
 If "Building dimension" is defined as a complex custom property, then building area, building height and building width are the multiple key value pairs defining the attribute building dimension.
```

[Screenshot to be provided]

> Fill all the mandatory fields and click on **Save** button to create a custom property. Either a Simple or a Complex custom property is created, based on inputs provided. In case of Complex custom property, each key value pair must be defined, and "type" should be provided for each key value pair.

Reviewer 2: ```suggestion
Fill all the mandatory fields and click **Save** to create a custom property. Based in the input, you create a either a **Simple custom property** or a **Complex custom property**. In case of a **Complex custom property** each key value pair must be defined, also provide the **Type** for each key value pair.
```

Reviewer 2: @MWindel Check placement of information after implementation

> 

Reviewer 2: @MWindel continue here

 If you select **Text**, the following fields appear:

*	**Min Length**:

Enter the minimum length of the text that must be provided for this custom property during the asset creation. Any text below the minimum length is not accepted.

If not selected, no minimum length criteria is set for this custom property during the asset creation.

* **Max Length**:

Enter the maximum length of the text that must be provided for this custom property during the asset creation. Any text above the maximum length is not accepted.

If the entered text exceeds the maximum length, you get notified of the incidence and the asset creation is cancelled.

* **RegExp**:

Enter a valid regular expression. During the asset creation, you must provide the custom property value, which adhers to the regular expression.

**Example:**

If the custom property is "Employee ID" with the regular expression "^SAGEMPID[a-zA-Z0-9]*$", all employee IDs have to start with the prefix “SAGEMPID”.

<a name="default-value"></a>
#### Default value

Assigns a default value to the custom property. If this field is left empty in the custom property, you must enter a value during the asset creation.

{{< c8y-admon-info>}}

> The default value has to adhere to all validations provided beforehand. If you have set the **Min length**, **Max length** and/or **RegExP**, the default value must fulfill all set validations.

Reviewer 2: Done

<a name=""></a>
#### File upload

This option allows you to upload a file to define the custom property.

If selected, enter the allowed file types separated by a comma in the **Allowed file types** field. When you create an asset with this custom property, you can only upload files of the allowed file types.

**Example:**

To upload the blueprint for a building, you must name the custom property "Building blueprint" and specify the file type of the blueprint file. Then upload the respective file. Once the asset is created, you can view this file for future reference.

<a name=""></a>
### To edit a custom property

1. Select the desired custom property in the **Property Library** to see all the details on the right of the screen.
2. Edit any content of the custom property.
3. Click **Save** to save your changes.

<a name=""></a>
### To search and filter custom properties

1. Navigate to the **Property Library** page.
2. Enter the desired name in the search field on the top left.
3. Click the search icon.

> All custom properties matching the search values are displayed on the screen on the right. If the search entry does not match with any custom properties, you see a notification.

Reviewer 2: Done

---
weight: 10
title: Property library
layout: redirect
---

> The **Property library** section allows you to create custom properties. Custom properties are the attributes that define an [asset type](/dtm/asset-types/#asset-types). One or more custom properties are used to define an asset type.

Reviewer: ```suggestion
The **Property library** section allows you to create custom properties. Custom properties are the attributes that define an [asset type](/dtm/asset-types/#asset-types). One or more custom properties define an asset type.
```

Also, the same custom property can belong to 2 different asset types.

### Custom property attributes

> When you navigate to the **Property library** page via **“Configuration> Digital twin library > Property library”** for the first time, there will be no custom property present in the system. Click **Add custom property** at the top right to add your first custom property.

Reviewer: ```suggestion
Navigate to **Configuration> Digital twin library > Property library** to open the **Property library** page. Initially, there are no custom properties available. Click **Add custom property** at the top right to add your first custom property.
```

> If you use a tenant which already has defined custom properties already defined **Property library**, you can find the list of defined custom properties on the left. The selected custom properties are listed on the right.

Reviewer: ```suggestion
If you use a tenant which already has custom properties defined, you find the list of defined custom properties on the left. The selected custom properties are listed on the right.
```

Author: Done

> #create-new-custom-property

Reviewer: Should this be an anchor?

### To create a new custom property

> In order to add a custom property, click **Add custom property** on the bottom left. Enter the required data in the dialog box on the right and click **Save**. You can now select this custom property.

Reviewer: In order to add > To create

![Property-add](/images/dtm/custom-property/dtm-property-library-create-property.png)

> Custom properties have the following attributes:

Reviewer: attributes > we usually talk about "parameters"; perhaps change for consistency

<table>
<col width="10">
<col width="90">
<thead>
<tr>
<th style="text-align:left">Field</th>
<th style="text-align:left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left"><b>Label</b></td>

> <td style="text-align:left">This information is mandatory. It refers to the name of the custom property. The custom property is further addressed using the information provided here.</td>

Reviewer: Please also add an extra column for "Mandatory"

</tr>
<tr>
<td style="text-align:left"><b>Key</b></td>
<td style="text-align:left">This information is mandatory. It is used to uniquely identify and store the custom property in the DTM.</td>
</tr>
<tr>
<td style="text-align:left"><b>Description</b></td>
<td style="text-align:left">This information is optional. It provides a brief description of the custom property.</td>
</tr>
<tr>
<td style="text-align:left"><b>Complex property toggle</b></td>
<td style="text-align:left">This information is optional. Use this toggle to create a complex custom property, that contains one or multiple key-value pairs. The toggle is disabled by default.</td>
</tr>
<tr>
<td style="text-align:left"><b>Type</b></td>

> <td style="text-align:left">This field defines the type of the custom property. It can be any of the following categories – Text, Number, File upload, Date picker, Enumeration or Boolean.</td>

Reviewer: Add a link to the detailed descriptions below

</tr>
<tr>
<td style="text-align:left"><b>Default value</b></td>
<td style="text-align:left">This information is optional. It defines the default value for the attribute. You can modify the content when creating an asset using this custom property.</td>
</tr>
</tbody>
</table>

{{< c8y-admon-info>}}

> Default value text box gets displayed only if the ‘Type’ chosen is either **Text** or **Number**.

Reviewer: ```suggestion
The **Default value** field is only shown if the **Type** chosen is either "Text" or "Number".
```
